src/pudl/transform/vceregen.py

```diff
@@ -0,0 +1,286 @@
+    )
+    df = pd.concat(
+        [df.reset_index(drop=True), new_time_col.reset_index(drop=True)], axis=1
+    ).rename(columns={"unnamed_0": "hour_of_year"})
+    return df
+
+
+def _stack_cap_fac_df(df: pd.DataFrame, df_name: str) -> pd.DataFrame:
+    """Funciton to transform each capacity factor table individually to save memory.
+
+    The main transforms are turning county columns into county rows and renaming columns
+    to be more human-readable and compatible with the FIPs df that will get merged in.
+
+    This function is intended to save memory by being applied to each individual
+    capacity factor table rather than the giant combined one.
+    """
+    logger.info(f"Stacking the county columns for {df_name} table.")
+    df_stacked = (
+        df.set_index(["hour_utc", "hour_of_year", "report_year"])
+        .stack()
+        .reset_index()
+        .rename(
+            columns={"level_3": "county_state_names", 0: f"capacity_factor_{df_name}"}
+        )
+        .assign(county_state_names=lambda x: pd.Categorical(x.county_state_names))
+    )
+    return df_stacked
+
+
+def _make_cap_fac_frac(df: pd.DataFrame, df_name: str) -> pd.DataFrame:
+    """Make the capacity factor column a fraction instead of a percentage.
+
+    This step happens before the table gets stacked to save memory.
+    """
+    logger.info(f"Converting capacity factor into a fraction for {df_name} table.")
+    county_cols = [x for x in df.columns if x not in ["report_year", "unnamed_0"]]
+    df[county_cols] = df[county_cols] / 100
+    return df
+
+
+def _check_for_valid_counties(
+    df: pd.DataFrame, clean_fips_df: pd.DataFrame, df_name: str
+) -> pd.DataFrame:
+    """Make sure the state_county values show up in the fips table.
+
+    This step happens before the table gets stacked to save memory.
+    """
+    logger.info(f"Checking for valid counties in the {df_name} table.")
+    county_state_names_fips = clean_fips_df.county_state_names.unique().tolist()
+    county_state_names_cap_fac = df.columns.tolist()
+    expected_non_counties = ["report_year", "unnamed_0"]
+    non_county_cols = [
+        x
+        for x in county_state_names_fips
+        if x not in county_state_names_cap_fac + expected_non_counties
+    ]
+    if non_county_cols:
+        raise AssertionError(f"""found unexpected columns that aren't in the FIPS table:
+{non_county_cols}.""")
+    return df
+
+
+def _combine_all_cap_fac_dfs(cap_fac_dict: dict[str, pd.DataFrame]) -> pd.DataFrame:
+    """Combine capacity factor tables."""
+    logger.info("Merging all the capacity factor tables into one")
+    merge_keys = ["report_year", "hour_utc", "hour_of_year", "county_state_names"]
+    for name, df in cap_fac_dict.items():
+        cap_fac_dict[name] = df.set_index(merge_keys)
+    mega_df = pd.concat(
+        [
+            cap_fac_dict["solar_pv"],
+            cap_fac_dict["offshore_wind"],
+            cap_fac_dict["onshore_wind"],
+        ],
+        axis=1,
```

Comment:
We've been trying to use the index/columns rather than 0/1 syntax for functions like concat since they're a bit less mysterious.

Reply:
Can you give me an example? I'm not sure what you mean by this @zaneselvans

Reply:
ah, yes this is way clearer! Thank you.